sync/pacman/pacman-utils

```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env bash
+
+wait_for_subshells_with_progress_and_output()
+{
+    # assign arguments to variables
+    subshell_pids=$1
+    FIFO=$2
+    
+    # get subshell counts
+    pid_count=${#subshell_pids[@]}
+    total_pids=$pid_count
+    
+    # add an empty newline so that the previous line isn't deleted
+    echo
+
+    while [ $pid_count -gt 0 ] || [ ! -z "$subshell_output" ] ; do
+        for (( i=0; i<$pid_count; i++ )); do
+            pid=${subshell_pids[$i]}
+            if ! kill -0 $pid 2> /dev/null; then
+                # process has finished, remove it from the array
+                # echo "process $pid has finished, remove it from the array"
+                unset subshell_pids[$i]
+            fi
+        done
+        subshell_pids=( ${subshell_pids[@]} )
+        pid_count=${#subshell_pids[@]}
+
+        # read the next line of output from the subprocesses from the $FIFO
+        # store the line of output in the $subshell_output variable
+        # `-t 1` to timeout if no output to read after 1 second
+        # `|| true` to prevent the script from exiting if read returns an error code (due to timeout)
+        read -t 1 subshell_output || true
+        
+        # if $subshell_output is not empty
+        if [ ! -z "${subshell_output}" ]; then  
+            # delete the previous line in the terminal
+            echo -en "\033[A\033[K"
+            echo "$subshell_output"
+            echo
+        fi
+
+        # calculate and print the progress
+        pids_remaining=$(( total_pids - pid_count ))
+        progress="$(( pids_remaining * 100 / total_pids ))% (${pids_remaining}/${total_pids})"
```

Comment:
For my personal taste, it could be like this to make more sense
`progress="$(( pids_remaining * 100 / total_pids ))%, pids remaining: (${pids_remaining}/${total_pids})"`
so that the output look something like:
`Downloading and extracting codebases: 89%, pids remaining: (53/59)`

Reply:
Updated